### PR TITLE
JAVA SDK templates: HTCL and Split

### PR DESCRIPTION
**Pull request by algonautshant**
ContractTemplate: main function is injecting parameter values.
HTCL and Split templates 
TestTemplates: tests both templates against the golden. HTCL is not matching the golden. Split is.

### CODE REVIEW

File: src/main/java/com/algorand/algosdk/templates/ContractTemplate.java [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+					}
+					progIdx += 31; // skip the parameter placeholder bytes in program bytecode
+					break;
+
+				case INT:
+					byte[] byteValue = putVarint(value.getIntValue());
+					for (byte b : byteValue) {
+						updatedProgram.add(b);	
+					}
+					progIdx += 0; // skip the parameter placeholder bytes in program bytecode
```
Comment (rotemh) [inline src/main/java/com/algorand/algosdk/templates/ContractTemplate.java]: nop?

File: src/main/java/com/algorand/algosdk/templates/ContractTemplate.java [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+		public int getIntValue() {
+			return intValue;
+		}
+
+		public byte[] getByteValue() {
+			return byteValue;
+		}
+	}
+
+	public String address;
```
Comment (rotemh) [inline src/main/java/com/algorand/algosdk/templates/ContractTemplate.java]: consider keeping the address as an address object

File: src/main/java/com/algorand/algosdk/templates/ContractTemplate.java [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+				case BASE64:
+					byte[] byteValueString = value.getByteValue();
+					byte[] byteValueLength = putVarint(byteValueString.length);
+					for (byte b : byteValueLength) {
+						updatedProgram.add(b);	
+					}
+					for (byte b : byteValueString) {
+						updatedProgram.add(b);	
+					}
+					progIdx += 1; // skip the parameter placeholder bytes in program bytecode
```
Comment (rotemh) [inline src/main/java/com/algorand/algosdk/templates/ContractTemplate.java]: Would be nice to comment the convention for the placeholder lengths

File: src/main/java/com/algorand/algosdk/templates/ContractTemplate.java [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+			if (b < 0x80) {
+				if (i > 9 || i == 9 && b > 1) {
+					return new int[] {0, -(i + 1)};
+				}
+				return new int[] {x | (b & 0xff) << s, i + 1};
+			}
+			x |= ((b & 0x7f) & 0xff) << s;
+			s += 7;
+		}
+		return new int[] {0,0};
```
[the remark this answers is not in the document]
Comment (algonautshant) [reply src/main/java/com/algorand/algosdk/templates/ContractTemplate.java]: It is used in the test, and it is public and can be used from elsewhere as well. 
I'll be happy to remove and remove the test if you prefer.

File: src/main/java/com/algorand/algosdk/templates/ContractTemplate.java [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+	 * @return ContractTemplate with the address and the program with the given parameter values
+	 * @throws NoSuchAlgorithmException
+	 */
+	protected static ContractTemplate inject(String program, int [] offsets, ParameterValue[] values) throws NoSuchAlgorithmException {
+
+		int paramIdx = 0;
+		if (offsets.length != values.length) {
+			throw new RuntimeException("offsets and values should have the same number of elements");
+		}
+		byte[] progBytes = Encoder.decodeFromBase64(program);
```
[the remark this answers is not in the document]
Comment (algonautshant) [reply src/main/java/com/algorand/algosdk/templates/ContractTemplate.java]: changed :-)

File: src/main/java/com/algorand/algosdk/templates/ContractTemplate.java [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+			return intValue;
+		}
+
+		public byte[] getByteValue() {
+			return byteValue;
+		}
+	}
+
+	public String address;
+	public String program;
```
[the remark this answers is not in the document]
Comment (algonautshant) [reply src/main/java/com/algorand/algosdk/templates/ContractTemplate.java]: changed :-)

File: src/main/java/com/algorand/algosdk/templates/ContractTemplate.java [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+	 * Each byte in a varint, except the last byte, has the most significant 
+	 * bit (msb) set – this indicates that there are further bytes to come. 
+	 * The lower 7 bits of each byte are used to store the two's complement 
+	 * representation of the number in groups of 7 bits, least significant 
+	 * group first.
+	 * https://developers.google.com/protocol-buffers/docs/encoding
+	 * @param value being serialized
+	 * @return byte array holding the serialized bits
+	 */
+	public static byte[] putVarint(int value) {
```
[the remark this answers is not in the document]
Comment (algonautshant) [reply src/main/java/com/algorand/algosdk/templates/ContractTemplate.java]: changed :-)

File: src/main/java/com/algorand/algosdk/templates/ContractTemplate.java [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+		private int intValue;
+
+		public ParameterValue(DataType dataType, String value) throws NoSuchAlgorithmException {
+			type = dataType;
+
+			if (type == DataType.ADDRESS) {
+				Address addr = new Address(value);
+				byteValue = addr.getBytes();
+			} else {
+				byteValue = value.getBytes();
```
[the remark this answers is not in the document]
Comment (algonautshant) [reply src/main/java/com/algorand/algosdk/templates/ContractTemplate.java]: changed :-)

File: src/main/java/com/algorand/algosdk/templates/ContractTemplate.java (hunk 1 above)
Comment (algonautshant) [reply src/main/java/com/algorand/algosdk/templates/ContractTemplate.java]: updated :-)

File: src/main/java/com/algorand/algosdk/templates/ContractTemplate.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (rotemh) [reply src/main/java/com/algorand/algosdk/templates/ContractTemplate.java]: that's fine, leave it

File: src/main/java/com/algorand/algosdk/templates/ContractTemplate.java (hunk 2 above)
Comment (algonautshant) [reply src/main/java/com/algorand/algosdk/templates/ContractTemplate.java]: changed! :-)

File: src/main/java/com/algorand/algosdk/templates/ContractTemplate.java [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+		private byte[] byteValue;
+		private int intValue;
+
+		public ParameterValue(DataType dataType, String value) throws NoSuchAlgorithmException {
+			type = dataType;
+
+			if (type == DataType.ADDRESS) {
+				Address addr = new Address(value);
+				byteValue = addr.getBytes();
+			} else {
```
Comment (rotemh) [inline src/main/java/com/algorand/algosdk/templates/ContractTemplate.java]: It will be better to explicitly check if it's `BASE64`

**Pull request merged**